### DISCUSSION TITLE
Some noob questions. Hopefully other noobs will read this thread in the future.

### DISCUSSION THREAD

**Discussion by toreupfeet** [mKeRix/room-assistant#552] [General]
I have Home Assistant installed via Docker on a Pi4B (running Debian 10). I have the Room Assistant add-on for HA. I am adding Pi Zero Ws around the house, which I believe creates a cluster. I want the Pi4B and each Pi Zero W to track the same BT Classic devices. I have some questions. I have spent time looking for answers on the RA site and in these threads, but I could use a little help from the experts.

1. Should the config be the exact same for every Pi? or...

2. Is there some extra yaml that I should add to one of the Pi's (I would prefer to add it to the Pi4B running HA) to make sure the Pi's are able to communicate with one-another properly?

3. Is there a way to streamline the configs so that there is one Pi that maintains the main config and updates the other Pi's? 
    - It seems that ansible is intended for this purpose. Perhaps I should look into that. 
    - Is ansible the only way this is possible?

4. Will the Pi Zero's performance be impacted if I add a gpio motion sensor (assuming a 6 second interval for BT Classic)?

5. Can the Pi4B 8gb 64bit (running HA) handle an interval of 1 for BT Classic? What would be an advisable interval?

6. Here is the `config/room_assistant/local.yaml` file for my HA add-on. How would I modify this to make it the default leader? What else should I add (I'll play with maxDistance on my own, but aside from that...)? And, what should the Pi Zero configs be, to make the cluster perform properly (aside from changing the instanceName)?

```
global:
  instanceName: debian
  integrations:
    - homeAssistant
    - bluetoothClassic
homeAssistant:
  mqttUrl: 'mqtt://192.168.50.192:1883'
  mqttOptions:
    username: homeassistant
    password: SuperSecretPassword
bluetoothClassic:
  interval: 6
  hciDeviceId:
    - 0
  addresses:
    - '00:00:00:00:00:00'
    - '00:00:00:00:00:00'    
    - '00:00:00:00:00:00'
```

I'm writing this because I have two Pi Zeros (so far), and one worked fine alongside the Pi4B, but when I added the second Pi Zero there appeared to be a conflict between the two devices in the log of Pi Zero number two. I don't recall what it was, specifically. It was a few days ago and I'm actually away from the house at the moment so I cannot check. I'd be happy to play with tweaks of the individual Pi Zeros on my own, as any good noob should, but I'd like to ensure I have the Pi's all communicating properly first. I'm sure other noobs out there will benefit from this thread.

**Comment by mKeRix** [comment]
Let me try to answer as best as I can:

1. Not necessarily - most stuff can stay the same across all Pis though. I would make sure that your Bluetooth settings are in sync across all for example, but the `instanceName` should be set individually to reflect wherever the device is located.
2. If everything is communicating fine for you there is no need for it, but you can still specify the [cluster settings](https://www.room-assistant.io/guide/cluster.html#settings) manually of course. `peerAddresses` are shared between the nodes during initial connections, so it's fine to add this to the instances causing trouble only. In your case it would be completely valid to just add that configuration block to the Pi 4B instance.
3. Currently you'd use ansible for that (you do need to still maintain the HA add-on config manually with that as of right now though!). There is an open issue requesting a more user-friendly option: #261.
4. GPIO shouldn't impact performance of the Pi.
5. The lower bound of BT Classic `interval` isn't limited by CPU power, but rather by the limitations of the Bluetooth protocol. You need some time for connections to be established, so setting the `scanTimeLimit` to something below 1 (always needs to be below `interval`) might cause detection to become flaky. On top of that, having constant connection requests running might have a really bad impact on your other 2.4 GHz wireless networks (WiFi, ZigBee...), as Bluetooth Classic connection attempts are quite aggressive on the frequency band. That being said, you can of course try it.
6. To make it the default leader take a look at [cluster usage tips](https://www.room-assistant.io/guide/cluster.html#usage-tips). Essentially you need to add a weight like below in each instance config, with your HA add-on having the highest number as weight.
```yaml
global:
  instanceName: debian
cluster:
  weight: 100
```

Other than that start with your config like that and see what it brings. Connectivity should be ensured automatically when possible, if you see trouble with the HA add-on you can take a look at what I wrote in point 2. Feel free to comment back here with any issues you run into!

**Comment by toreupfeet** [comment]
Thank you the quick reply! Heiko, we're all very fortunate to have such a remarkable presence detection solution. I'm juuuuust savvy enough to be able how to figure out how to use this sort of tech, and to think there are people like you who are not only able to _create_ it, but that you're also so involved in the community... I feel very blessed, and humbled. Thank you.

The cluster seems to be working. 
I haven't experimented with the intervals, but here are my configs for now:

Pi4B running HA (config/room_assistant/local.yaml)
```
global:
  instanceName: Debian
  autoDiscovery: true
  integrations:
    - homeAssistant
    - bluetoothClassic
bluetoothClassic:
  interval: 4
  scanTimeLimit: 3
  hciDeviceId:
    - 0
  addresses:
    - '00:00:00:00:00:00'
    - '00:00:00:00:00:00'
    - '00:00:00:00:00:00'
cluster:
  weight: 100
```

And here are my Pi Zero W configs
```
global:
  instanceName: pi_zero_front_door #unique to each Pi Zero W
  integrations:
    - homeAssistant
    - bluetoothClassic
homeAssistant:
  mqttUrl: 'mqtt://192.168.50.192:1883'
  mqttOptions:
    username: homeassistant
    password: SuperSecretPassword
bluetoothClassic:
  interval: 6
  scanTimeLimit: 5
  preserveState: false
  hciDeviceId:
    - 0
  addresses:
    - '00:00:00:00:00:00'
    - '00:00:00:00:00:00'
    - '00:00:00:00:00:00'
```

For any future noobs trying to build a cluster, perhaps start with a similar config. 

I ran into an issue (not sure how it started) but this `pi_zero_front_door` Pi Zero actually isn't working quite right. It did work, but after some tinkering the service was somehow masked. I unmasked it, and now the Pi is sluggish and RA works only when I run `room-assistant` from Putty. But it does work as a cluster when I do so, so I am optimistic that the configs are correct. I'm going to reinstall tonight and try again (and add a third Pi Zero as well). The kitchen Pi Zero works great. It tends to be the leader despite my having given the Pi4B a weight of 100, but I'm not very concerned.

Thank you again.
